---
title: "EC_TCGA_vs_gnomAD"
author: "Olga Kondrashova"
date: "July 23, 2020"
output:
  html_document:
    includes:
      df_print: paged
      toc: true
      toc_float: true
      toc_depth: 3
---

```{r getPackageFunction, echo=FALSE, eval=TRUE, message=FALSE, warning=FALSE}
# Load required packages
getPackage <- function(pkg, load = TRUE, silent = TRUE, 
                       repos = "http://cran.us.r-project.org") {
  if(!suppressMessages(suppressWarnings(require(pkg, character.only = TRUE,
                                                quietly = silent)))) {
    install.packages(pkg, repos = repos, quiet = TRUE)
    source("http://bioconductor.org/biocLite.R")
    biocLite(pkg)
  }
  if(load) 
    suppressPackageStartupMessages(library(pkg, character.only = TRUE,
                                           quietly = silent))
  return("i")
}

x <- c("tidyverse","magrittr","knitr","data.table","here","formatR","kableExtra",
       "deconstructSigs","MutationalPatterns","cowplot","circlize","ComplexHeatmap","stringr")

invisible(lapply(x, getPackage))
rm(x)
```


```{r setup000, eval=TRUE, include=FALSE}
options(scipen=999)

# knitr global options
knitr::opts_chunk$set(root.dir = here(),comment = NA,cache = TRUE,echo = TRUE, message = FALSE,warnings = FALSE, 
                      include = TRUE,fig.align = 'center', figcap.prefix = "Figure",figcap.sep = ":", 
                      figcap.prefix.highlight = "**",fig.path = "figures/", 
                      tidy.opts = list(width.cutoff = 80), tidy = TRUE)

```

# Filtering gnomAD variants by consequence

This analysis is done as part of the case-control endometrial cancer study in collaboration with Mandy Spurdle.   Repeating Jannah Shamsani's PhD thesis experiment with version 2.1 of gnomAD data.  

Filtering will include:  
* Predicted truncating variants^  
* Pathogenic/likely pathogenic based on ClinVar assertions.

^Predicted truncating variants are nonsense, frameshift indels, native splice donor or acceptor sites (dinucleotide consensus region).  


First read in RefSeq transcript ids to use (used the latest versions - as determined by VEP annotations). 

```{r supporting_info}
df_refseq_ids <- fread("data/supporting/genomic_coordinates_30genes_ensemble_hg19.csv")


manual_vars <- scan("data/supporting/Variants_to_include_manual_annotation_2020Aug.txt", character())

trunc_filter <- "splice_acceptor_variant|splice_donor_variant|frameshift_variant|stop_gained"
noncoding_filter <- c("3_prime_UTR_variant","5_prime_UTR_variant",
                      "downstream_gene_variant","upstream_gene_variant")


```

```{r filter_functions}
filter_trunc_variants <- function(df){
  df %>%
    filter(!SYMBOL %in% c("POLE","POLD1")) %>%
    filter((SYMBOL == "PMS2" & !((Pos_start > 6010505 & Pos_start < 6027301) | 
                                   (Pos_start > 6031553 & Pos_start < 6031738))) |
             SYMBOL != "PMS2" ) %>% # Manual filter, was too difficult to do with dataframe
    # select fs and ns (exclude last exon) and splice site variants only
    filter((str_detect(Consequence,
                       "frameshift_variant|stop_gained") & Last_exon != "yes") |
             str_detect(Consequence,"splice_acceptor_variant|splice_donor_variant")) %>%  
    #has to not include "benign" in the clin sig. description
    filter(!str_detect(ClinVar_CLNSIG,
                       fixed("benign", ignore_case = T))) 
}

filter_clinvar_variants <- function(df, noncoding_filter = noncoding_filter){
  df %>% 
    filter(!SYMBOL %in% c("POLE","POLD1")) %>%
    filter((SYMBOL == "PMS2" & !((Pos_start > 6010505 & Pos_start < 6027301) | 
                                   (Pos_start > 6031553 & Pos_start < 6031738))) |
             SYMBOL != "PMS2" ) %>% # Manual filter, was too difficult to do with dataframe
    # exclude non-coding region variants
    filter(!Consequence %in%  noncoding_filter) %>% 
    #has to contain "pathogenic" in the clin sig. description
    #and not include "conflicting" in the clin sig. description
    #and exclude variants that have "no_assertion_criteria_provided"
    #or include variants that have been manually curated
    filter((str_detect(ClinVar_CLNSIG,
                      fixed("pathogenic", ignore_case = T)) & 
             !str_detect(ClinVar_CLNSIG,
                         fixed("conflicting", ignore_case = T)) & 
             !str_detect(ClinVar_CLNREVSTAT,
                         "no_assertion_criteria_provided")) | 
             HGVSc %in% manual_vars)
}

filter_non_clinvar <- function(df,  genes = "MMR"){
  if (genes == "MMR") {
    df %<>% 
      filter(SYMBOL %in% c("PMS2","MLH1","MSH2","MSH6")) %>%
      filter((SYMBOL == "PMS2" & !((Pos_start > 6010505 & Pos_start < 6027301) | 
                                     (Pos_start > 6031553 & Pos_start < 6031738))) |
               SYMBOL != "PMS2" ) 
  } else if (genes == "MUTYH"){
    df %<>% 
      filter(SYMBOL == "MUTYH") 
  }
  df %>% 
    # exclude non-coding region variants
    filter( str_detect(Consequence,"missense_variant")) %>% 
    #has to contain "pathogenic" in the clin sig. description
    #and not include "conflicting" in the clin sig. description
    #and exclude variants that have "no_assertion_criteria_provided"
    filter(ClinVar_CLNSIG == "-") 
}

filter_POLE_POLD1_variants <- function(df){
  df %>% 
    filter(SYMBOL %in% c("POLE","POLD1")) %>%
    # include only missense variants
    filter(str_detect(Consequence,"missense_variant")) %>%
    #has to contain "pathogenic" in the clin sig. description
    #and not include "conflicting" in the clin sig. description
    #and exclude variants that have "no_assertion_criteria_provided"
    filter(str_detect(ClinVar_CLNSIG,
                      fixed("pathogenic", ignore_case = T)) & 
             !str_detect(ClinVar_CLNSIG,
                         fixed("conflicting", ignore_case = T)) & 
             !str_detect(ClinVar_CLNREVSTAT,
                         "no_assertion_criteria_provided")) 
}

```

## gnomAD exomes

Read in VEP annotated gnomAD exome variants for selected 30 genes.  
Pre-filtered already by RefSeq IDs using python script *filter_VEP_step1.py*.  
(Filter by relevant RefSeq transcripts, preferred transcript in ClinVar - taken from Jannah Shamsani's thesis) 
```{r read_exomes, message=FALSE, warning=FALSE}

df_exomes <- fread("data/gnomad/gnomad.exomes.r2.1.1.genes30.vep.refseq.tsv.gz",
                   skip="#Uploaded_variation",
                   check.names = TRUE) %>%
mutate_at(.vars=c("gnomADe_non_cancer_AC_nfe","gnomADe_non_cancer_AN_nfe",
                 "gnomADe_non_cancer_nhomalt_nfe","gnomADg_AC_nfe",
                 "gnomADg_AN_nfe","gnomADg_nhomalt_nfe"),
                 funs(as.numeric(.)))
  
```

```{r add_columns_exomes, message=FALSE}

df_exomes %<>%
  separate(Location, into = c("Pos_chr","Pos_coord"), sep = "[:]", remove=FALSE) %>%
  separate(Pos_coord, into = c("Pos_start"), sep = "-", extra = "drop") %>%
  separate(EXON, into=c("EXON_Var","EXON_Count"),sep = "/", remove = FALSE, fill="right") %>%
  mutate(Last_exon = if_else(EXON_Var == EXON_Count ,"yes", "no"))

```

Select all predicted truncating - nonsense & frameshift indels (exclude last exon), native splice donor or acceptor sites (dinucleotide consensus region).  
Remove variants not found in nfe non-cancer population (excludes TCGA).  

```{r filter_exomes_trunc}

df_exomes_trunc <- df_exomes %>%
  filter(gnomADe_non_cancer_AC_nfe !=0) %>%
  filter_trunc_variants(.)

```
   
Then select ClinVar variants - likely pathogenic or pathogenic.  
Only taking variants without conflicting information & some level of assertion.  
Remove variants not found in nfe non-cancer population (excludes TCGA).  

```{r filter_exomes_clinvar}

df_exomes_clinvar <- df_exomes %>%
  filter(gnomADe_non_cancer_AC_nfe !=0) %>% 
  filter_clinvar_variants(., noncoding_filter)

```

Applying a separate filter for POLE & POLD1 variants as interested only in missense variants

```{r filter_exomes_POLE_POLD1}

df_exomes_POLE_POLD1 <- df_exomes %>%
  filter(gnomADe_non_cancer_AC_nfe !=0) %>% 
  filter_POLE_POLD1_variants(.)

```

As control number of alleles using max allele number per gene (as Jannah used in PhD thesis) before filtering by variants of interest variants.

```{r calculate_exomes_total_AN}

df_exomes_AN <- df_exomes %>%
  group_by(SYMBOL) %>%
  summarise(median_AN_nfe = median(gnomADe_non_cancer_AN_nfe),
            max_AN_nfe = max(gnomADe_non_cancer_AN_nfe))

```

Combine truncating and ClinVar classified variants.  

```{r exomes_combine_variants}

#combine truncating and clinvar classified variants
df_exomes_merged <- bind_rows(df_exomes_trunc,df_exomes_clinvar, df_exomes_POLE_POLD1) %>%
  distinct() %>%
  #group by gene
  group_by(SYMBOL) %>%
  summarise(
    #sum up the alternative allele count 
    sum_AC_nfe = sum(gnomADe_non_cancer_AC_nfe), 
    # add up all homozygous carriers (individuals, not alleles)
    sum_hom_nfe = sum(gnomADe_non_cancer_nhomalt_nfe)) %>% 
  full_join(df_exomes_AN, by = "SYMBOL") %>%
  replace(is.na(.),0) %>%
  mutate(
    #count total number of carriers 
    #(remove one of hom. allele, so that hom. carriers are not counted twice)
    carrier_nfe = sum_AC_nfe-sum_hom_nfe, 
    #calculate total number of individuals 
    #(need to round down as used max above, so may not divide cleanly)
    total_nfe = as.integer(max_AN_nfe/2),
    #calculate the allele frequency
    freq_AF_nfe = round(sum_AC_nfe/max_AN_nfe,4),
    #calculate frequency of carriers
    freq_carrier_nfe = round(carrier_nfe/total_nfe,4))
```

## gnomAD genomes

Read in VEP annotated gnomAD genome variants for selected 30 genes.  
Pre-filtered already by refseq ids using python script *filter_VEP_step1.py*.  

```{r read_genomes, cache.lazy = FALSE}
df_genomes <-fread("data/gnomad/gnomad.genomes.r2.1.1.genes30.vep.refseq.tsv.gz",
                    skip="#Uploaded_variation",
                   check.names = TRUE) %>%
  mutate_at(.vars=c("gnomADe_non_cancer_AC_nfe","gnomADe_non_cancer_AN_nfe",
                    "gnomADe_non_cancer_nhomalt_nfe","gnomADg_AC_nfe",
                    "gnomADg_AN_nfe","gnomADg_nhomalt_nfe"),
            funs(as.numeric(.)))

```

```{r add_columns_genomes, message=FALSE}

df_genomes %<>%
  separate(Location, into = c("Pos_chr","Pos_coord"), sep = "[:]", remove=FALSE) %>%
  separate(Pos_coord, into = c("Pos_start"), sep = "-", extra = "drop") %>%
  separate(EXON, into=c("EXON_Var","EXON_Count"),sep = "/", remove = FALSE, fill="right") %>%
  mutate(Last_exon = if_else(EXON_Var == EXON_Count ,"yes", "no"))

```

First select all presumed truncating - nonsense & frameshift indels (exclude last exon), native splice donor or acceptor sites (dinucleotide consensus region).  
Remove variants not found in nfe population.  

```{r filter_genomes_trunc}

df_genomes_trunc <- df_genomes %>% 
  filter(gnomADg_AC_nfe !=0) %>%
  filter_trunc_variants(.)

```

Then select ClinVar variants - likely pathogenic or pathogenic.  
Only taking variants without conflicting information & some level of assertion.  
Remove variants not found in nfe population.  

```{r filter_genomes_clinvar}
df_genomes_clinvar <- df_genomes %>% 
  filter(gnomADg_AC_nfe !=0) %>%
  filter_clinvar_variants(.,noncoding_filter)
```

Separate filter for POLE & POLD1 variants.

```{r filter_genomes_POLE_POLD1}
df_genomes_POLE_POLD1 <- df_genomes %>%
  filter(gnomADg_AC_nfe !=0) %>% 
  filter_POLE_POLD1_variants(.)

```

As described in **gnomAD exome** section, decided to take a max allele (also calculate median) number per gene before filtering by variants of interest variants (but after filtering by transcript ID).  

```{r calculate_genomes_total_AN}
df_genomes_AN <- df_genomes %>%
  group_by(SYMBOL) %>%
  summarise(median_AN_nfe = median(gnomADg_AN_nfe),
            max_AN_nfe = max(gnomADg_AN_nfe))
```
  
Combine truncating and ClinVar classified variants.  

```{r genomes_combine_variants}
#combine truncating and clinvar classified variants
df_genomes_merged <- rbind(df_genomes_trunc,df_genomes_clinvar,df_genomes_POLE_POLD1) %>% 
  distinct() %>%
  group_by(SYMBOL) %>% #group by gene
  summarise(
    #sum up the alternative allele count 
    sum_AC_nfe = sum(gnomADg_AC_nfe), 
    # add up all homozygous carriers (individuals, not alleles)
    sum_hom_nfe = sum(gnomADg_nhomalt_nfe)) %>%
  full_join(df_genomes_AN, by = "SYMBOL") %>%
  replace(is.na(.),0) %>%
  mutate(
    #count total number of carriers
    #(remove one of hom. allele, so that hom. carriers are not counted twice)
    carrier_nfe = sum_AC_nfe-sum_hom_nfe, 
    #calculate total number of individuals 
    #(need to round down as used median above, so may not divide cleanly)
    total_nfe = as.integer(max_AN_nfe/2),
    #calculate the allele frequency
    freq_AF_nfe = round(sum_AC_nfe/max_AN_nfe,4),
    #calculate frequency of carriers
    freq_carrier_nfe = round(carrier_nfe/total_nfe,4)) 
```
  
## Combining gnomAD exome and genome tables

Sum up all values and recalculate frequencies.  

```{r combine_exome_and_genome}

df_gnomad_merged <- df_exomes_merged %>%
  full_join(df_genomes_merged, by = "SYMBOL",suffix = c("_e","_g")) %>%
 # select(-contains("freq")) %>%
  mutate(sum_AC_nfe = sum_AC_nfe_e + sum_AC_nfe_g,
         sum_hom_nfe = sum_hom_nfe_e + sum_hom_nfe_g,
         sum_AN_nfe = max_AN_nfe_e + max_AN_nfe_g,
         carrier_nfe = carrier_nfe_e + carrier_nfe_g,
         total_nfe = total_nfe_e + total_nfe_g,
         #calculate the allele frequency
         freq_AF_nfe = round(sum_AC_nfe/sum_AN_nfe,4), 
         #calculate frequency of carriers
         freq_carrier_nfe = round(carrier_nfe/total_nfe,4)) 

df_gnomad_merged %>%
  write_tsv(.,"./tables/gnomAD_total_counts_and_freq.tsv")

cat("Final combined (exome and genome) frequency table\n")
kable(df_gnomad_merged) %>% 
  kable_styling(bootstrap_options = c("striped", "hover", "condensed", "responsive"))

```


## Saving ClinVar variant lists for manual review (combined exome and genome).
```{r save_variants_gnomAD, warning=FALSE}

# Save all ClinVar variants selected by current path/likely path criteria
bind_rows(df_exomes_trunc,df_exomes_clinvar, df_exomes_POLE_POLD1) %>%
  distinct() %>%
  select(HGVSc,SYMBOL,Consequence,
         gnomADe_non_cancer_AN_nfe,gnomADe_non_cancer_AC_nfe,
         gnomADg_AN_nfe,gnomADg_AC_nfe,
         contains("clinvar", ignore.case = T)) %>%
  full_join(bind_rows(df_genomes_trunc,df_genomes_clinvar,df_genomes_POLE_POLD1) %>% 
              distinct() %>%
              select(HGVSc,SYMBOL,Consequence,
                     gnomADe_non_cancer_AN_nfe,gnomADe_non_cancer_AC_nfe,
                     gnomADg_AN_nfe,gnomADg_AC_nfe,
                     contains("clinvar", ignore.case = T))) %>%
  distinct() %>%
  write_tsv(.,"./tables/gnomAD_combined_variants.tsv")

# Save path/likely path ClinVar variants with no assertion
df_exomes %>%
  filter((SYMBOL == "PMS2" & !((Pos_start > 6010505 & Pos_start < 6027301) | 
                                 (Pos_start > 6031553 & Pos_start < 6031738))) |
           SYMBOL != "PMS2" ) %>%  
  select(HGVSc,SYMBOL,Consequence,
         gnomADe_non_cancer_AN_nfe,gnomADe_non_cancer_AC_nfe,
         gnomADg_AN_nfe,gnomADg_AC_nfe,
         contains("clinvar", ignore.case = T)) %>%
  full_join(df_genomes %>%
              filter((SYMBOL == "PMS2" & !((Pos_start > 6010505 & Pos_start < 6027301) | 
                                             (Pos_start > 6031553 & Pos_start < 6031738))) |
                       SYMBOL != "PMS2" ) %>%  
              select(HGVSc,SYMBOL,Consequence,
                     gnomADe_non_cancer_AN_nfe,gnomADe_non_cancer_AC_nfe,
                     gnomADg_AN_nfe,gnomADg_AC_nfe,
                     contains("clinvar", ignore.case = T))) %>%
  filter(gnomADe_non_cancer_AC_nfe !=0 & gnomADg_AC_nfe !=0) %>%  
  # exclude fs, ns and splice site variants
  filter(!(str_detect(Consequence, trunc_filter))) %>% 
  #remove all variants outside of coding region
  filter(!Consequence %in% noncoding_filter)  %>% 
  #has to contain "pathogenic" in the clin sig. description
  filter(str_detect(ClinVar_CLNSIGA,fixed("pathogenic", ignore_case = T)) & 
           #has to not include "conflicting" in the clin sig. description
           !str_detect(ClinVar_CLNSIG,fixed("conflicting", ignore_case = T)) & 
           #select only variants that have "no_assertion_criteria_provided"
           str_detect(ClinVar_CLNREVSTAT,"no_assertion_criteria_provided"))  %>% 
  write_tsv(.,"./tables/gnomAD_combined_ClinVar_no_assertion.tsv")

# Save path/likely path ClinVar variants with conflicting evidence
df_exomes %>% 
  filter((SYMBOL == "PMS2" & !((Pos_start > 6010505 & Pos_start < 6027301) | 
                                 (Pos_start > 6031553 & Pos_start < 6031738))) |
           SYMBOL != "PMS2" ) %>%  
  select(HGVSc,SYMBOL,Consequence,
         gnomADe_non_cancer_AN_nfe,gnomADe_non_cancer_AC_nfe,
         gnomADg_AN_nfe,gnomADg_AC_nfe,
         contains("clinvar", ignore.case = T)) %>%
  full_join(df_genomes %>%
              filter((SYMBOL == "PMS2" & !((Pos_start > 6010505 & Pos_start < 6027301) | 
                                             (Pos_start > 6031553 & Pos_start < 6031738))) |
                       SYMBOL != "PMS2" ) %>%  
              select(HGVSc,SYMBOL,Consequence,
                     gnomADe_non_cancer_AN_nfe,gnomADe_non_cancer_AC_nfe,
                     gnomADg_AN_nfe,gnomADg_AC_nfe,
                     contains("clinvar", ignore.case = T))) %>%
  filter(gnomADe_non_cancer_AC_nfe !=0 & gnomADg_AC_nfe !=0) %>%  
  # exclude fs, ns and splice site variants
  filter(!(str_detect(Consequence,trunc_filter))) %>% 
  #remove all variants outside of coding region
  filter(!Consequence %in% noncoding_filter)  %>% 
  #has to contain "pathogenic" in the clin sig. 
  filter(str_detect(ClinVar_CLNSIGA,"Pathogenic|Likely_pathogenic") & 
           #include only "conflicting" in the clin sig. description
           str_detect(ClinVar_CLNSIG, 
                      fixed("conflicting", ignore_case = T))) %>% 
  write_tsv(.,"./tables/gnomAD_combined_ClinVar_conflicting.tsv")

```

## TCGA UCEC cases

This data cannot be publicly shared because it requires ethics clearance.  
  
Read in VEP annotated TCGA UCEC variants for selected 30 genes.   
Filter by relevant RefSeq transcripts.
```{r read_TCGA, message=FALSE, cache.lazy=FALSE}
df_TCGA_sample_codes <- read_tsv("data/supporting/TCGA_UCEC_donor_info.txt")


df_TCGA <- list.files(path="data/TCGA/",
                          pattern = ".vep.refseq.tsv.gz$",
                          recursive = F,
                          full.names = T) %>%
  purrr::set_names(.) %>% 
  map_dfr(~fread(.,skip="#Uploaded_variation",
                   check.names = TRUE, colClasses = "character"), .id = "source") %>%
  mutate(file_name = str_remove(basename(source),".vep.refseq.tsv.gz")) %>%
  mutate(file_name = str_replace(file_name,"_vs_",".")) %>%
  separate(file_name, 
           into = c("testDonorLabel","testSampleLabel","testBam",
                    "case_id","aliquot_id","controlBam"), sep="[.]") %>%
  left_join(df_TCGA_sample_codes) 
```

```{r add_columns_TCGA, message=FALSE}

df_TCGA %<>%
  separate(Location, into = c("Pos_chr","Pos_coord"), sep = "[:]", remove=FALSE) %>%
  separate(Pos_coord, into = c("Pos_start"), sep = "-", extra = "drop") %>%
  separate(EXON, into=c("EXON_Var","EXON_Count"),sep = "/", remove = FALSE, fill="right") %>%
  mutate(Last_exon = if_else(EXON_Var == EXON_Count ,"yes", "no"))

```

Filter variants as in gnomAD sections.
```{r filter_TCGA_trunc}

df_TCGA_trunc <- df_TCGA %>%
  filter_trunc_variants(.)

```

```{r filter_TCGA_clinvar}

df_TCGA_clinvar <- df_TCGA %>%
  filter_clinvar_variants(.,noncoding_filter)

```

```{r filter_TCGA_POLE_POLD1}

df_TCGA_POLE_POLD1 <- df_TCGA %>%
  filter_POLE_POLD1_variants(.) 

```

Combine truncating and ClinVar classified variants.  

```{r TCGA_combine_variants, message=FALSE}
#combine truncating and clinvar classified variants
df_TCGA_merged_temp <- rbind(df_TCGA_trunc,df_TCGA_clinvar,df_TCGA_POLE_POLD1) %>% 
  distinct() %>%
  group_by(SYMBOL)#group by gene


# Access germline MAF and check allele count by pulling out number of reads supporting variant and alt reads
df_TCGA_check_AC <- df_TCGA_merged_temp %>%
  # Change file name to ghcc maf file
  mutate(source = str_replace(source,".vep.refseq.tsv",".ghcc.maf")) %>%
  mutate(source = str_replace(source,"data/TCGA",
                              "/working/genomeinfo/share/analysis/TCGA_-_UCEC/MAFs")) %>%
  pull(source) %>%
  purrr::set_names(.) %>% 
  #Read individual mafs
  map_dfr(~fread(.,skip="Hugo_Symbol\t",check.names = TRUE, colClasses = "character"),
          .id="source") %>%
  # Change file name back for merging dataframes later
  mutate(source = str_replace(source,".ghcc.maf",".vep.refseq.tsv")) %>%
  mutate(source = str_replace(source,
                              "/working/genomeinfo/share/analysis/TCGA_-_UCEC/MAFs",
                              "data/TCGA")) %>%
  dplyr::rename(SYMBOL = Hugo_Symbol,
                Pos_start = Start_Position) %>%
  select(source,SYMBOL, Chromosome, Pos_start, CDS_Change,
         N_Depth, N_Ref_Count, N_Alt_Count) %>%
  # Pull out only variatns in the selected list
  right_join(df_TCGA_merged_temp %>%
               select(SYMBOL, Pos_chr, Pos_start, Allele, source)) %>% 
  #Calculate VAF, AC and hom_counts
  mutate(VAF = as.numeric(N_Alt_Count)/as.numeric(N_Depth)) %>%
  mutate(AC_count = if_else(VAF<0.7,1,2)) %>%
  mutate(carrier_count=1,
         hom_count = if_else(AC_count == 2,1,0)) 

```

``` {r filter_TCGA_eu_cases}
df_TCGA_EU_ethnicity <- df_TCGA_sample_codes %>%
  filter(SNP_ethnicity == "European" | 
           (SNP_ethnicity == "#N/A" & race =="white")) 

df_TCGA_EU_ethnicity %>%
  select(testDonorPublicationID, project_id, race, SNP_ethnicity, primary_diagnosis) %>%
  mutate(SNP_ethnicity = str_replace(SNP_ethnicity, "#N/A","not available")) %>%
  write_tsv("./tables/TCGA_UCEC_eu_cases.tsv")

df_TCGA_merged <- df_TCGA_merged_temp %>%
  left_join(df_TCGA_check_AC) %>%
  filter(testDonorPublicationID %in% df_TCGA_EU_ethnicity$testDonorPublicationID) %>%
  select(SYMBOL, HGVSc,Consequence, testDonorPublicationID,
         carrier_count, AC_count, hom_count) %>%
  distinct() %>%
  group_by(SYMBOL) %>%
  summarise(
    #sum up carrier count
    carrier_eu = sum(carrier_count),
    sum_AC_eu = sum(AC_count),
    sum_hom_eu = sum(hom_count)) %>%
  mutate(
    # total number of individuals 
    total_eu = 308,
    sum_AN_eu = total_eu*2,
    #calculate frequency of carriers
    freq_carrier_eu = round(carrier_eu/total_eu,4),
    freq_AF_eu = round(sum_AC_eu/sum_AN_eu,4)) 
```

Final carrier frequencies table for gnomAD genomes:  
```{r TCGA_final_table}
kable(df_TCGA_merged) %>% 
  kable_styling(bootstrap_options = c("striped", "hover", "condensed", "responsive"))

df_TCGA_merged %>%
  write_tsv(.,"./tables/TCGA_total_counts_and_freq.tsv")

```

```{r TCGA_gnomAD_final_table}
df_gnomad_merged %>%
  left_join(df_TCGA_merged) %>%
  select(SYMBOL,carrier_eu, sum_hom_eu, total_eu, freq_carrier_eu,carrier_nfe,sum_hom_nfe,total_nfe,freq_carrier_nfe) %>%
  dplyr::rename(Gene = SYMBOL, 
                N_carrier_TCGA = carrier_eu, N_hom_TCGA =sum_hom_eu, 
                N_total_TCGA = total_eu, Freq_TCGA = freq_carrier_eu,
                N_carrier_gnomAD = carrier_nfe, N_hom_gnomAD =sum_hom_nfe, 
                N_total_gnomAD = total_nfe, Freq_gnomAD = freq_carrier_nfe) %>%
  mutate(Freq_TCGA = Freq_TCGA*100, Freq_gnomAD = Freq_gnomAD*100) %>%
  arrange(-N_carrier_TCGA,-Freq_TCGA,-N_carrier_gnomAD, Gene) %>%
  replace(is.na(.),0) %>%
  write_tsv(.,"./tables/TCGA_gnomAD_total_counts_and_freq.tsv")

```

```{r TCGA_save_variants}

# Save all ClinVar variants selected by current path/likely path criteria
df_TCGA_merged_variants <- df_TCGA_merged_temp %>% 
  select(HGVSc,SYMBOL,Pos_chr,Pos_start,Allele, aliquot_id, Consequence, testDonorPublicationID,
         contains("clinvar", ignore.case = T)) %>%
  distinct() %>%
  group_by(HGVSc,SYMBOL,Consequence) %>%
  mutate(TCGA_cases = paste0(testDonorPublicationID, collapse=";"),
         aliquot_id = paste0(aliquot_id, collapse = ";")) %>%
  select(-testDonorPublicationID, aliquot_id) %>%
  distinct() %>%
  write_tsv(.,"./tables/TCGA_variants.tsv")

# Same but for EU cases
df_TCGA_merged_variants_eu <- df_TCGA_merged_temp %>% 
  filter(testDonorPublicationID %in% df_TCGA_EU_ethnicity$testDonorPublicationID) %>%
  select(HGVSc,SYMBOL,Consequence, testDonorPublicationID,
         contains("clinvar", ignore.case = T)) %>%
  distinct() %>%
  group_by(HGVSc,SYMBOL,Consequence) %>%
  mutate(TCGA_eu_cases = paste0(testDonorPublicationID, collapse=";")) %>%
  select(-testDonorPublicationID) %>%
  distinct() %>%
  write_tsv(.,"./tables/TCGA_eu_variants.tsv")


# Save path/likely path ClinVar variants with no assertion
df_TCGA %>%
  filter((SYMBOL == "PMS2" & !((Pos_start > 6010505 & Pos_start < 6027301) | 
                                 (Pos_start > 6031553 & Pos_start < 6031738))) |
           SYMBOL != "PMS2" ) %>%  
  select(HGVSc,SYMBOL,Consequence,testDonorPublicationID,
         contains("clinvar", ignore.case = T)) %>%
  # exclude fs, ns and splice site variants
  filter(!(str_detect(Consequence, trunc_filter))) %>% 
  #remove all variants outside of coding region
  filter(!Consequence %in% noncoding_filter)  %>% 
  #has to contain "pathogenic" in the clin sig. description
  filter(str_detect(ClinVar_CLNSIGA,fixed("pathogenic", ignore_case = T)) & 
           #has to not include "conflicting" in the clin sig. description
           !str_detect(ClinVar_CLNSIG,fixed("conflicting", ignore_case = T)) & 
           #select only variants that have "no_assertion_criteria_provided"
           str_detect(ClinVar_CLNREVSTAT,"no_assertion_criteria_provided"))  %>% 
    group_by(HGVSc,SYMBOL,Consequence) %>%
  mutate(case_id = paste0(testDonorPublicationID, collapse=";")) %>%
  select(-testDonorPublicationID) %>%
  distinct() %>%
  write_tsv(.,"./tables/TCGA_ClinVar_no_assertion.tsv")

# Save path/likely path ClinVar variants with conflicting evidence
df_TCGA %>% 
  filter((SYMBOL == "PMS2" & !((Pos_start > 6010505 & Pos_start < 6027301) | 
                                 (Pos_start > 6031553 & Pos_start < 6031738))) |
           SYMBOL != "PMS2" ) %>%  
  select(HGVSc,SYMBOL,Consequence, testDonorPublicationID,
         contains("clinvar", ignore.case = T)) %>%
  # exclude fs, ns and splice site variants
  filter(!(str_detect(Consequence,trunc_filter))) %>% 
  #remove all variants outside of coding region
  filter(!Consequence %in% noncoding_filter)  %>% 
  #has to contain "pathogenic" in the clin sig. 
  filter(str_detect(ClinVar_CLNSIGA,
                    "Likely_pathogenic|Pathogenic") & 
           #include only "conflicting" in the clin sig. description
           str_detect(ClinVar_CLNSIG,
                      fixed("conflicting", ignore_case = T))) %>% 
    group_by(HGVSc,SYMBOL,Consequence) %>%
  mutate(case_id = paste0(testDonorPublicationID, collapse=";")) %>%
  select(-testDonorPublicationID) %>%
  distinct() %>%
  write_tsv(.,"./tables/TCGA_ClinVar_conflicting_tsv")

```

 Supp table - variants list
```{r save_gnomAD_TCGA_variants}

# Save all ClinVar variants selected by current path/likely path criteria
bind_rows(df_exomes_trunc,df_exomes_clinvar, df_exomes_POLE_POLD1) %>%
  distinct() %>%
  select(HGVSc,SYMBOL,Consequence,
         gnomADe_non_cancer_AC_nfe,gnomADe_non_cancer_nhomalt_nfe,
         gnomADg_AC_nfe, gnomADg_nhomalt_nfe) %>%
  full_join(bind_rows(df_genomes_trunc,df_genomes_clinvar,df_genomes_POLE_POLD1) %>% 
              distinct() %>%
              select(HGVSc,SYMBOL,Consequence,
                     gnomADe_non_cancer_AC_nfe,gnomADe_non_cancer_nhomalt_nfe,
                     gnomADg_AC_nfe, gnomADg_nhomalt_nfe)) %>%
  distinct() %>%
  replace(is.na(.),0) %>%
  mutate(gnomAD_nfe_exome = gnomADe_non_cancer_AC_nfe -gnomADe_non_cancer_nhomalt_nfe,
         gnomAD_nfe_genome = gnomADg_AC_nfe-gnomADg_nhomalt_nfe) %>%
  full_join(df_TCGA_merged_variants %>%
              select(HGVSc, SYMBOL, Consequence, TCGA_cases)) %>%
  full_join(df_TCGA_merged_variants_eu %>%
              select(HGVSc, SYMBOL, Consequence, TCGA_eu_cases)) %>%
  replace(is.na(.),0) %>%
  select(-c(gnomADe_non_cancer_AC_nfe, gnomADe_non_cancer_nhomalt_nfe, gnomADg_AC_nfe, gnomADg_nhomalt_nfe)) %>%
  arrange(SYMBOL, HGVSc) %>%
  write_tsv(.,"./tables/TCGA_gnomAD_all_variants.tsv")
    


```

---

# Somatic mutational signature analysis

This section is looking at the mutational signatures in cases with germline variants.


## TCGA Signatures
```{r read_supporting_files}
vars_list <-  df_TCGA_merged_variants %>%
  ungroup() %>%
  separate_rows(aliquot_id, TCGA_cases, sep=";") %>%
  select(HGVSc, SYMBOL, Pos_chr, Pos_start, Allele, aliquot_id, TCGA_cases) %>%
  mutate_at(c("aliquot_id","TCGA_cases"), ~str_trim(.))
```

Read in all MAF files.

```{r read_maf_som}
filenames <- list.files(path="./data/MAFs",
                        pattern = ".shc.maf.gz$",
                        recursive = F,
                        full.names = T) 


maf_shc <- data.frame()
for (i in filenames){
  sample_id <- strsplit(basename(i), split = "[.]")[[1]][4] #normal sample ID
  if (sample_id %in% vars_list$aliquot_id){
    maf_shc <- rbind(maf_shc,
                     read_tsv(i, comment = "#", col_types = cols(.default = "c")))
  }     
}

maf_shc %<>%
  separate(Tumor_Sample_Barcode, into = c("case_submitter_id","aliquot_id"),
           sep=":", remove = FALSE, extra = "drop") %>%
  filter(!aliquot_id %in% c("5b259ecb-d09b-40d4-9f46-3585ab178e52","851870d9-cd58-42d8-9cb4-f785b62aa0bd")) # filter 2 samples that have double tumour/normal pairs. 

```

```{r read_maf_germ}
filenames <- list.files(path="./data/MAFs",
                        pattern = ".ghcc.maf.gz$",
                        recursive = F,
                        full.names = T) 


maf_ghcc <- data.frame()
for (i in filenames){
  sample_id <- strsplit(basename(i), split = "[.]")[[1]][4] #normal sample ID
  if (sample_id %in% vars_list$aliquot_id){
    maf_ghcc %<>%
      bind_rows(read_tsv(i, comment = "#", col_types = cols(.default = "c")))
  }     
}

maf_ghcc %<>%
  separate(Tumor_Sample_Barcode, into = c("case_submitter_id","aliquot_id"),
           sep=":", remove = FALSE, extra = "drop") %>%
  filter(!aliquot_id %in% c("5b259ecb-d09b-40d4-9f46-3585ab178e52","851870d9-cd58-42d8-9cb4-f785b62aa0bd")) %>%
  # filter 2 samples that have double tumour/normal pairs. 
  separate(Matched_Norm_Sample_Barcode, into = c("case_submitter_id","normal_aliquot_id"),
           sep=":", remove = FALSE, extra = "drop")

sample_list <- vars_list %>%
  left_join(maf_ghcc %>% select(case_submitter_id, normal_aliquot_id) %>% distinct(),
            by = c("aliquot_id" = "normal_aliquot_id")) %>%
  select(case_submitter_id,TCGA_cases)
``` 
 
 Read in msi sensor outputs.
```{r read_msi}

msi_data <- list.files(path="./data/msi_sensor",
                         full.names = T) %>%
  purrr::set_names(. %>% basename()) %>% 
  map_dfr(~read_tsv(.,col_types = cols()), .id = "name") %>%
  separate(name,c("case_submitter_id","aliquot_id"),sep="_",remove=FALSE) %>%
  filter(aliquot_id %in% maf_shc$aliquot_id) %>%
  left_join(sample_list)

```

Read in library capture information.
```{r read_meta}
capture_info <- read_tsv("./data/supporting/TCGA_-_endometrial_cancer.StudyPairedAnalysis.tsv",
                         col_types = cols())  %>%
  select(testSampleLabel, testCaptureKit)

```


Define *deconstructSigs* and *combine_plot* functions.
```{r define_func}
run_deconstructsigs <- function(data,out_path,out_name,signatures, method = "default", plot=FALSE, cutoff=0.15) {
  # Run deconstructSigs on all samples
  for (i in 1:nrow(data)){
    sample_id=row.names(data[i,])
    sigs= whichSignatures(tumor.ref = data, 
                          signatures.ref = signatures, 
                          sample.id = sample_id,
                          signature.cutoff = cutoff,
                          contexts.needed = TRUE,
                          tri.counts.method = method)
    # Plot signatures if defined
    if ( isTRUE(plot)){
      pdf(paste0(out_path,"figures/",sample_id,"_",method,"_",out_name,".pdf"))
      plotSignatures(sigs)
      dev.off()
    }
    # Combine sig assignments for samples to one dataframe
    if (exists("sigs_final")) { 
      sigs_final <- rbind(sigs_final,sigs$weights) 
    } else { 
      sigs_final <- sigs$weights
    }
  }
  # Subset only detected signatures
  sigs_detected <- sigs_final %>%
    rownames_to_column(var="Sample")%>%
    gather(Sig, Count, -Sample) %>%
    group_by(Sig) %>%
    filter(Count > 0) %>%
    spread(Sig, Count, fill = 0)
  # Return signature assigment dataframe
  write.table(sigs_detected, paste0(out_path,"tables/",method,"_",out_name,".tsv"),quote = F, sep = "\t", row.names = F)
  return(sigs_detected)
}


combined_plot <- function(gvar,sig,sig2 = NULL,tmb,svar_POLE,msi,sample_order,
                          group_genes = NULL,sig_type = "pcawg", sig_type2 = "denovo") {
  if(sig_type == "cosmicv2"){
    sig_colours <- sig_colours_cosmicv2
  } else if (sig_type == "denovo" ){
    sig_colours <- sig_colours_denovo
  } else if (sig_type == "pcawg"){
    sig_colours <- sig_colours
  }
  if(sig_type2 == "cosmicv2"){
    sig_colours2 <- sig_colours_cosmicv2
  } else if (sig_type2 == "denovo" ){
    sig_colours2 <- sig_colours_denovo
  } else if (sig_type2 == "pcawg"){
    sig_colours2 <- sig_colours
  }
  
  if (is_tibble(group_genes)){
    p1 <- gvar %>%
      left_join(group_genes, by ="Gene") %>%
      filter(Sample %in% sample_order) %>%
      mutate(Sample = factor(Sample, levels = sample_order)) %>%
      mutate(Gene = factor(Gene, levels = group_genes$Gene)) %>%
      ggplot(.) +
      geom_tile(aes(y=Gene,x=Sample,fill=Type, colour = Enriched, width = 0.8, height = 0.8), size = 1) +
      facet_grid(Group~., scales="free_y", space = "free_y")
    
  } else{
  p1 <- gvar %>%
    filter(Sample %in% sample_order) %>%
    mutate(Sample = factor(Sample, levels = sample_order)) %>%
    ggplot(.) +
    geom_tile(aes(y=Gene,x=Sample,fill=Type, colour = Enriched, width = 0.8, height = 0.8), size = 1)
  }
  p1 <- p1 +
    scale_fill_manual(values = c("WT"="#d6d6d6",
                                 "Pathogenic"="#1e6dc0")) + 
    #scale_fill_manual(values= mut_colours) +
    scale_colour_manual(values = c("yes"="#E8294A","white")) +
    guides(color = guide_legend(ncol = 1,
                                override.aes = list(fill=NA))) +
    theme_minimal(base_size = 11) + labs(x = "", y = "", colour = "Enriched in tumour",
                                         fill = "Germline Status") +
    theme(axis.text.x=element_blank(),
          panel.grid.major.y = element_blank(),
          panel.grid.major.x = element_blank(),
          legend.position = "right",
          strip.text.y = element_blank())

  p2 <- sig %>%
    filter(Sample %in% sample_order) %>%
    mutate(Sample = factor(Sample, levels = sample_order)) %>%
    ggplot(.) +
    geom_bar(aes(x= Sample, y = Count, fill = Sig_full), stat = "identity", position = "stack") + 
    labs(x="",y="Proportion",fill ="Signature (Aetiology)") +
    theme_minimal(base_size = 11) +
    scale_fill_manual(values = sig_colours) +
    guides(fill = guide_legend(ncol = 1)) +
    theme(axis.text.x=element_blank(),
          strip.text.x = element_text(size = 6),
          panel.grid.minor = element_blank(),
          #panel.grid.major.y = element_blank(),
          panel.grid.major.x = element_blank(),
          legend.position = "right")
  
  p3 <- tmb %>%
    filter(TCGA_cases %in% sample_order) %>%
    mutate(Sample = factor(TCGA_cases, levels = sample_order)) %>%
    ggplot(.)+
    geom_bar(aes(x= Sample, y = TMB, fill = "TMB"), stat = "identity") + 
    labs(x="",y="Mutations per Mb",fill = "") +
    scale_y_log10()+
    theme_minimal(base_size = 11) +
    scale_fill_manual(values = "grey") +
    theme(axis.text.x=element_blank(),
          strip.text.x = element_text(size = 6),
          panel.grid.minor = element_blank(),
          #panel.grid.major.y = element_blank(),
          panel.grid.major.x = element_blank(),
          legend.position = "right")
  
  p4 <- svar_POLE %>%
    filter(TCGA_cases %in% sample_order) %>%
    mutate(Sample = factor(TCGA_cases, levels = sample_order)) %>%
    ggplot(.) +
    geom_tile(aes(y=Gene,x=Sample,fill=Variant_Classification, width = 0.8, height = 0.8), size = 1, colour="white") +
    scale_fill_manual(values= c("WT"="#d6d6d6",
                                "Mutation"="#1e6dc0",
                                 "Hom. Deletion" = "#C92E12",
                                 "Methylation" = "#2FD9C9")) +
    guides(color = guide_legend(ncol = 1)) +
    theme_minimal(base_size = 11) + labs(x = "", y = "", fill = "Somatic Status") +
    theme(axis.text.x=element_text(angle = 90, hjust = 1 , vjust = 0.5),
          panel.grid.minor = element_blank(),
          panel.grid.major.y = element_blank(),
          #panel.grid.major.x = element_blank(),
          legend.position = "right")
  
  p5 <- msi %>%
    filter(TCGA_cases %in% sample_order) %>%
    mutate(Sample = factor(TCGA_cases, levels = sample_order)) %>%
    mutate(MSI_threshold = if_else(`%`>=3.5,">=3.5","<3.5")) %>%
    ggplot(.)+
    geom_bar(aes(x=Sample,y=`%`, fill = MSI_threshold),stat="identity")+
    #geom_hline(yintercept=3.5,colour="black",linetype="dashed")+
    theme_minimal(base_size = 11) + labs(x = "", y = "MSI score (%)", fill = "MSIsensor threshold") +
    theme(axis.text.x=element_blank(), 
          panel.grid.minor = element_blank(),
          #panel.grid.major.y = element_blank(),
          panel.grid.major.x = element_blank()
          )
  
  if (!is.null(sig2)){
    p6 <- sig2 %>%
      filter(Sample %in% sample_order) %>%
      mutate(Sample = factor(Sample, levels = sample_order)) %>%
      ggplot(.) +
      geom_bar(aes(x= Sample, y = Count, fill = Sig_full), stat = "identity", position = "stack") + 
      labs(x="",y="Proportion",fill ="De novo Signature (Aetiology)") +
      theme_minimal(base_size = 11) +
      scale_fill_manual(values = sig_colours2) +
      guides(fill = guide_legend(ncol = 1)) +
      theme(axis.text.x=element_blank(),
            strip.text.x = element_text(size = 6),
            panel.grid.minor = element_blank(),
            #panel.grid.major.y = element_blank(),
            panel.grid.major.x = element_blank(),
            legend.position = "right")
    p_main <- plot_grid(p3, p5, p2, p6, p1, p4, align = "v",axis = 'lr', nrow = 6, rel_heights = c(1/6, 1/6, 1/3,  1/3, 1/3, 1/4))
  } else {
    p_main <- plot_grid(p3, p5, p2, p1, p4, align = "v",axis = 'lr', nrow = 5, rel_heights = c(1/6, 1/6, 1/3, 1/3, 1/4))
  }
  
  return(p_main)
}

```

Other supporting info.
```{r define_supp_info}
#Chromosome order
chr_levels=c("1","2","3","4","5","6","7","8","9","10","11",
             "12","13","14","15","16","17","18","19","20","21",
             "22","X","Y")


sig_list_cosmicv2 <- c("Sig18 (MUTYH mut)",
                       "Sig6 (MSI)","Sig15 (MSI)","Sig21 (MSI)", "Sig26 (MSI)",
                       "Sig20 (MSI)", "Sig14 (POLE mut & MSI)",
                       "Sig10 (POLE mut)",  
                       "Sig3 (HRD)", "Sig30 (NTHL1 mut)", 
                       "Sig13 (APOBEC)","Sig2 (APOBEC)","Sig1 (Age)",
                       "Other")

sig_colours_cosmicv2 <- c("#c5ff1a",
                          "#8d69a3","#7a5093", "#541e75","#420666",
                          "#1f09ba","#fa939e",
                          "#f97886",
                          "#943503","#118866",
                          "#3498db","#57cef2","#009999", 
                          "#8d958f")
names(sig_colours_cosmicv2) <- sig_list_cosmicv2

sig_list_denovo <- c("Sig D (MSI, Sig26)",
                     "Sig C (MSI, Sig20)","Sig E (POLE mut & MSI, Sig14)",
                     "Sig A (POLE mut, Sig10)",
                     "Sig F (APOBEC, Sig2)",
                     "Sig B (Age, Sig1)")

sig_colours_denovo<- c("#420666",
                       "#1f09ba","#fa939e",
                       "#f97886",
                       "#3498db",
                       "#009999")

names(sig_colours_denovo) <- sig_list_denovo



mut_colours <- c("#d6d6d6","#90CAF9","#FFAB91","#CE93D8","#80DEEA",
                "#E6EE9C","#EC769F")

names(mut_colours) <- c("WT", "Nonsense", "Frame_Shift", "Missense", "Splice_Site", 
                        "In-frame_indel", "Frame_Shift;Missense")


p96_colours <- c("#30B3EA","#000000","#E40000","#989797","#93C700","#F3ADBB")
names(p96_colours) <- c("C>A","C>G","C>T","T>A","T>C","T>G")

col_fun_cosine <- colorRamp2(c(0, 0.5, 1), c("#F7F9DD","#4DA563","#022D50"))

```


## Tumour Mutation Burden analysis

```{r tmb_analysis}
tmb_df <- maf_shc %>% 
  left_join(capture_info, by = c("aliquot_id" = "testSampleLabel")) %>%
  left_join(sample_list) %>%
  count(TCGA_cases, case_submitter_id, testCaptureKit) %>%
  mutate(captureSize = case_when(testCaptureKit == "hg18 nimblegen exome version 2" ~ 26.2,
                                 testCaptureKit == "Human All Exon 44Mb (SureSelect)" ~ 44,
                                 testCaptureKit == "SeqCap EZ Human Exome Library v2.0 (Nimblegen)" ~ 36.5,
                                 testCaptureKit == "SeqCap EZ Human Exome Library v3.0 (Nimblegen)" ~ 64)) %>%
  mutate(TMB = round(n/captureSize,2)) %>%
  arrange(-TMB) 

```

## Subset by mutations

```{r svar}

som_variants <- maf_shc %>% 
  left_join(capture_info, by = c("aliquot_id" = "testSampleLabel")) %>%
  left_join(sample_list) %>%
  filter(Hugo_Symbol %in% df_refseq_ids$Gene) %>%
  mutate(T_Alt_Count = as.integer(T_Alt_Count), T_Depth = as.integer(T_Depth))

som_variants %>%
write_tsv("./tables/somatic_variants.tsv")


```

## DeconstructSigs Mutational Signatures

Subset mafs by SNVs only, annotate and format for deconstructSigs.  
Filter samples with >= 100 variants.  
```{r subset_snv}
maf_shc_snv <- maf_shc %>% 
  left_join(sample_list) %>%
  filter(Variant_Type == "SNP") %>%
  mutate(alt = ifelse(Reference_Allele == Tumor_Seq_Allele1, Tumor_Seq_Allele2, Tumor_Seq_Allele1)) %>%
  filter(Chromosome %in% chr_levels) %>%
  mutate(chr = paste0("chr",Chromosome)) %>%
  group_by(TCGA_cases) %>%
  filter(n() >=100) %>%
  ungroup() %>%
  mutate(Start_Position = as.integer(Start_Position)) %>%
  select(TCGA_cases, chr, Start_Position, Reference_Allele, alt) %>%
  data.frame() # deconstructSigs does not like tibbles

sigs_input <- mut.to.sigs.input(mut.ref = maf_shc_snv, 
                                sample.id = "TCGA_cases", 
                                chr = "chr", 
                                pos = "Start_Position", 
                                ref = "Reference_Allele", 
                                alt = "alt")
```  

Read in Cosmic v2 signatures.
```{r cosmic_sigs, warning=FALSE, message = FALSE}
cosmic_t <- read_delim("./data/supporting/signatures_probabilities.txt",
                      delim = "\t") 

cosmic_t %<>%
  select(1:33,-`Substitution Type`,-Trinucleotide) %>%
  arrange(match(`Somatic Mutation Type`,colnames(sigs_input))) %>%
  column_to_rownames(var = "Somatic Mutation Type") %>%
  as.matrix()

cosmic_sigs <- as.data.frame(t(cosmic_t)) %>%
  rownames_to_column(var = "Signature") %>%
  mutate(Signature = str_replace(Signature,"Signature ","Sig")) %>%
  column_to_rownames(var = "Signature")


```

Convert to deconstructSigs matrix & run deconstructSigs (in exome mode).  
Using 0.15 signature cutoff.
```{r run_dSigs}

sig_exome_cosmic <- run_deconstructsigs(data = sigs_input,
                                        out_path = "./",
                                        out_name = "cosmicv2_sig_",
                                        signatures = cosmic_sigs, 
                                        method = "exome", 
                                        plot = F,
                                        cutoff = 0.15)

```

```{r parse_dSigs_out}

sig_exome_cosmic_melt <- sig_exome_cosmic %>%
  mutate(Other = (1- rowSums(.[-1]))) %>% #Calculate up to 100%
  gather(Sig, Count, -Sample) %>%
  filter(Count != 0) %>%
  mutate(Sig_name = case_when(Sig == "Sig1" ~ "Age",
                              Sig == "Sig10" ~ "POLE mut",
                              Sig == "Sig13" ~ "APOBEC",
                              Sig == "Sig2" ~ "APOBEC",
                              Sig == "Sig14" ~ "POLE mut & MSI",
                              Sig == "Sig15" ~ "MSI",
                              Sig == "Sig20" ~ "MSI",
                              Sig == "Sig21" ~ "MSI",
                              Sig == "Sig26" ~ "MSI",
                              Sig == "Sig3" ~ "HRD",
                              Sig == "Sig30" ~ "NTHL1 mut",
                              Sig == "Sig6" ~ "MSI",
                              Sig == "Other" ~ "",
                              Sig == "Sig18" ~ "MUTYH mut")) %>%
  mutate(Sig_full = if_else(Sig == "Other", Sig, paste0(Sig," (",Sig_name,")")))

sig_exome_cosmic_melt %<>%
  mutate(Sig_full = factor(Sig_full, levels = rev(sig_list_cosmicv2)))

sig_exome_cosmic_melt %>%
  select(1:3) %>%
  arrange(Sample,-Count) %>%
  write_tsv(paste0("./tables/cosmicv2_sig_exome_melt.tsv"))

```

Create somatic variant subsets for POLE, MMR and other gene variants.  
  
For POLE variant subset including all missense variants in exons 9-14 (except for c.1204T>C).  
Variant c.1204T>C is not reported in COSMIC, but reported in 6 alleles in gnomAD. More importantly, sample TCGA_UCEC_325 that has this variant does not have increase mutational burden or POLE signature - the only som POLE sample without signature.  
Manually found c.890C>T (p.S297F) hotspot mutation in TCGA_UCEC_064 that was filtered out due to low coverage.
  
```{r subset_svar}
# Subset POLE variants only
som_variants_POLE<- som_variants %>%
  filter(Hugo_Symbol == "POLE", Variant_Classification == "Missense_Mutation") %>%
  filter(Exon_Intron_Rank %in% 9:14) %>%
  filter(!CDS_Change == "c.1204T>C")  %>% 
  mutate(VAF = paste0(round(T_Alt_Count/T_Depth*100,0),"%")) %>%
  mutate(Mutation = paste(Transcript_ID,CDS_Change, Amino_Acid_Change,VAF)) %>%
  select(TCGA_cases, Variant_Classification, Mutation) %>%
  right_join(select(sig_exome_cosmic, Sample), by=c("TCGA_cases"="Sample")) %>%
  replace_na(list(Variant_Classification = "WT")) %>%
  mutate(Gene= "POLE") %>%
  mutate(Mutation = case_when(TCGA_cases == "TCGA_UCEC_064" ~ "ENST00000320574 c.890C>T p.S297F 40%",
                              TRUE ~ Mutation),
         Variant_Classification = case_when(TCGA_cases == "TCGA_UCEC_064" ~ "Missense_Mutation",
                                            TRUE ~ Variant_Classification)) %>%
  mutate(Variant_Classification = str_remove(Variant_Classification,"_Mutation"))


# Subset non-POLE variants only
som_variants_filt <- som_variants %>%
  filter(Hugo_Symbol != "POLE") %>%
  filter(grepl("Frame|Missense|Nonsense|Splice",Variant_Classification))%>%
  mutate(VAF = paste0(round(T_Alt_Count/T_Depth*100,0),"%")) %>%
  mutate(Alt = if_else(Reference_Allele == Tumor_Seq_Allele1,Tumor_Seq_Allele2,Tumor_Seq_Allele1)) %>%
  mutate(Mutation = if_else(CDS_Change!="null",paste(Hugo_Symbol,Transcript_ID,CDS_Change, Amino_Acid_Change,VAF),
                       paste0(Hugo_Symbol," ",Variant_Classification," ",
                              Chromosome,":",Start_Position,Reference_Allele,">",Alt, " ",VAF))) %>%
  select(Hugo_Symbol,TCGA_cases,Mutation)

# Subset MMR variants only
som_variants_MMR <- som_variants_filt %>%
  filter(Hugo_Symbol %in% c("MLH1","MSH2","MSH6","PMS2")) %>%
  select(-Hugo_Symbol)

som_variants_MMR %>%
  write_tsv("./tables/TCGA_somatic_MMR_variants_all.tsv")

#Save variant_list
somatic_variants_added <- vars_list %>%
  left_join(som_variants_POLE %>% 
              select(TCGA_cases,Mutation),
            by = c("TCGA_cases")) %>%
  group_by_at(vars(-Mutation)) %>%
  summarise(POLE_som_mut = paste(Mutation, collapse = "; ")) %>%
  ungroup() %>%
  left_join(som_variants_filt, by = c("TCGA_cases", "SYMBOL" = "Hugo_Symbol")) %>%
  group_by_at(vars(-Mutation)) %>%
  summarise(Som_mut_gene_of_interest = paste(Mutation, collapse = "; ")) %>%
  ungroup() %>%
  left_join(som_variants_MMR, by = c("TCGA_cases")) %>%
  group_by_at(vars(-Mutation)) %>%
  summarise(Som_mut_MMR = paste(Mutation, collapse = "; ")) %>%
  write_tsv("./tables/table_somatic_variants_added.tsv")

```

```{r prepare_gvar}

germline_vars <- vars_list %>%
  separate(HGVSc, into = c("RefSeq_ID","CDS_Change"), sep = ":", remove = FALSE) %>%
  mutate(Chromosome = str_remove(Pos_chr,"chr"),
         Start_Position = as.character(Pos_start)) %>%
  left_join(maf_ghcc %>%
              select(-aliquot_id), 
            by = c("Chromosome","Start_Position", "aliquot_id" = "normal_aliquot_id")) %>%
  mutate(Enriched = if_else(as.integer(T_Alt_Count)/as.integer(T_Depth) > 0.6, "yes","no")) %>%
  mutate(Variant_Class = case_when(str_detect(Variant_Classification,"RNA") ~ "Splice_Site",
                                   str_detect(Variant_Classification,"Frame_Shift") ~ "Frame_Shift",
                                   str_detect(Variant_Classification,"_Mutation") ~
                                     str_remove(Variant_Classification,"_Mutation"),
                                   str_detect(Variant_Classification,"In_Frame") ~"In-frame_indel",
                                   TRUE ~ Variant_Classification)) %>%
  select(Hugo_Symbol, TCGA_cases,Enriched,Variant_Class) %>%
  mutate(Var = paste(Variant_Class,Enriched,sep=";")) %>%
  mutate(Var = str_remove(Var,";no")) %>%
  select(-Enriched, -Variant_Class) %>%
  group_by(Hugo_Symbol,TCGA_cases) %>%
  summarise(Var = paste0(Var, collapse = ";")) %>%
  spread(key = Hugo_Symbol, value = Var, fill = "WT") %>%
  gather(key = Gene, value = Type, -TCGA_cases) %>%
  mutate(Enriched = if_else(str_detect(Type,";yes"),"yes","no"),
         Type = str_remove(Type,";yes")) %>%
  dplyr::rename(Sample = TCGA_cases) %>%
  mutate(Type = if_else(Type == "WT", Type, "Pathogenic")) %>%
  filter(Sample %in% sig_exome_cosmic_melt$Sample)

```

Adding MLH1 methylation information downloaded from cBioPortal. Decided to used beta-value 0.3 cutoff for methylated vs unmethylated.  
Also, downloaded CNA information for MSH2 from cBioPortal.
```{r load_MLH1me_MSH2del_MMR_mut, message=FALSE}

MLH1me <- read_tsv("data/other/MLH1_Methylation_HM27.txt") %>%
  inner_join(read_tsv("data/other/MLH1_Methylation_HM450.txt"))%>%
  mutate(Variant_Classification = case_when(`MLH1: Methylation (HM450)` > 0.3 ~ "Methylation",
                          `MLH1: Methylation (HM27)` > 0.3 ~ "Methylation",
                          TRUE ~ "WT")) %>%
  mutate(Gene = "MLH1") %>%
  dplyr::rename(submitter_id = `Patient ID`) %>%
  left_join(df_TCGA_sample_codes %>%
              select(testDonorPublicationID, submitter_id))  %>%
  dplyr::rename(TCGA_cases=testDonorPublicationID)

MLH1me %>%
 ggplot(aes(x=`MLH1: Methylation (HM27)`)) +
 geom_histogram(bins = 50) +
 geom_vline(xintercept = 0.3)

MLH1me %>%
 ggplot(aes(x=`MLH1: Methylation (HM450)`)) +
 geom_histogram(bins = 50) +
 geom_vline(xintercept = 0.3)

MSH2_del <- read_tsv("data/other/cna_MSH2.txt") %>%
  mutate(Variant_Classification = case_when(MSH2 == "-2" ~ "Hom. Deletion",
                          TRUE ~ "WT")) %>%
  mutate(submitter_id = str_remove(SAMPLE_ID, "-01$|-02$")) %>%
  mutate(Gene = "MSH2")  %>%
    left_join(df_TCGA_sample_codes %>%
              select(testDonorPublicationID, submitter_id))  %>%
  dplyr::rename(TCGA_cases=testDonorPublicationID)

#Manually annotated variants with Clinvar, left only truncating and LP and P (removed PMS2 in psedogene region)
MMR_muts_som <- read_tsv("data/TCGA/TCGA_somatic_MMR_variants_all_annotated.txt") %>%
  filter(Include == "yes") %>%
  separate(Mutation, into = c("Gene"), sep = "[ ]",extra = "drop") %>%
  mutate(Variant_Classification = "Mutation") %>%
  select(-Include)

som_info_POLE_MMR <- som_variants_POLE %>%
  bind_rows(MSH2_del %>%
               filter(TCGA_cases %in% som_variants_POLE$TCGA_cases & 
                        Variant_Classification != "WT") %>%
               select(TCGA_cases, Variant_Classification, Gene)) %>%
  bind_rows(MLH1me %>%
              filter(TCGA_cases %in% som_variants_POLE$TCGA_cases & 
                        Variant_Classification != "WT") %>%
              select(TCGA_cases, Variant_Classification, Gene)) %>%
  bind_rows(MMR_muts_som) %>%
  select(-Mutation) %>%
  group_by(TCGA_cases, Gene) %>%
  distinct() %>%
  mutate(Variant_Classification = str_replace(Variant_Classification,"Missense","Mutation")) %>%
  spread(key =TCGA_cases, value = Variant_Classification) %>%
  replace(is.na(.),"WT") %>%
  gather(key = TCGA_cases, value = Variant_Classification, -Gene)

```



## Prepare combined plots with mutational signatures

```{r combined_plot_path_subset_cosmicv2, message=FALSE, fig.height=12,fig.width=20}

 sig_genes<- tibble(Gene = rev(c("MSH2","MSH6","PMS2",
                                 "BRCA1","PALB2","RAD51C","FAN1","CHEK2","BRIP1","NBN",
                                 "NTHL1","MUTYH")),
                    Group = rev(c("MMR","MMR","MMR",
                                  "HR","HR","HR","HR","HR","HR","HR",
                                  "NTHL1","MUTYH"))) %>%
  mutate(Group = factor(Group, levels=rev(unique(Group)))) %>%
  mutate(Gene = factor(Gene, levels = rev(Gene)))

germ_vars_path <- germline_vars %>%
  filter(Gene %in% sig_genes$Gene) %>%
  group_by(Sample) %>%
  filter(!all(Type == "WT")) %>%
  ungroup() %>%
  mutate(Gene = factor(Gene, levels = rev(sig_genes$Gene))) %>%
  arrange(Type,Gene) 


combined_plot(
  gvar = germ_vars_path,
  sig = sig_exome_cosmic_melt,
  tmb = tmb_df,
  svar_POLE = som_info_POLE_MMR,
  msi = msi_data,
  sample_order = unique(germ_vars_path$Sample),
  group_genes = sig_genes,
  sig_type = "cosmicv2"
)

ggsave("./figures/combined_plot_path_subset_cosmicv2.pdf",height = 12, width = 20)

```

---

# Somatic MUTYH sample (FFPE)

Read in FFPE sample trinucleotide counts (done by Felicity Newell).  
```{r read_ffpe}
ffpe_input_t <- read_tsv("./data/other/FFPE_sample_trinucleotide_counts.tsv",
                       col_types = cols()) %>%
  mutate(sample = "FFPE") %>%
  spread(Trinucleotide.context, FFPE_sample) %>%
  column_to_rownames("sample")
```
  
Assess mutational signature distribution using dSigs against pcawg. 
```{r run_dSigs_ffpe}

sig_ffpe <- run_deconstructsigs(data = ffpe_input_t, 
                                out_path = "./",
                                out_name = "cosmicv2_sig_",
                                signatures = cosmic_sigs,
                                method = "default",
                                plot = T)

```
  
Plot 96 trinucleotide profile for the FFPE sample.

```{r plot96context_ffpe, fig.width=8,fig.height=3}
ffpe_input_t  %>%
  rownames_to_column("Sample") %>%
  gather("Pos","Count", -Sample) %>%
  separate(Pos, into = c("Start","End"), sep = ">", remove = FALSE) %>%
  separate(Start, into = c("First","Second")) %>%
  separate(End, into = c("Third","Forth")) %>%
  mutate(Context = paste0(First, Second,Forth), Pos = paste0(Second,">",Third)) %>%
  select(-c(First,Second,Third,Forth)) %>%
  ggplot(.) +
  geom_bar(aes(x=Context,y=Count, fill = Pos), stat = "identity") +
  scale_fill_manual(values = p96_colours) +
  scale_y_continuous(labels = scales::number_format(accuracy = 1)) +
  facet_grid(.~Pos, scales = "free_x", space = "free_x") +
  labs(title = "MUTYH variant - tumour sample") +
  theme(axis.text.x = element_text(angle = 90, hjust = 1, size = 6,family="mono"),
        legend.position = "none",
        axis.title = element_text(size = 10),
        axis.text.y = element_text(size = 6),
        strip.text = element_text(size = 10, face = "bold"),
        panel.grid.major.y = element_line(colour = "grey80")) 

ggsave("./figures/plot96context_ffpe.pdf", width = 8, height = 3)
```

```{r plot_sig_proportion_FFPE_default,fig.width=3,fig.height=5}
sig_ffpe %>%
  mutate(Other = (1- rowSums(.[-1]))) %>% #Calculate up to 100%
  gather(Sig, Count, -Sample) %>%
  filter(Count != 0) %>%
  mutate(Sig_name = case_when(Sig == "Other" ~ "",
                              Sig == "Sig18" ~ "MUTYH mut")) %>%
  mutate(Sig_full = if_else(Sig == "Other", Sig, paste0(Sig," (",Sig_name,")"))) %>%
  ggplot(.) +
  geom_bar(aes(x= Sample, y = Count, fill = Sig_full), stat = "identity", position = "stack") + 
  labs(x="",y="Proportion",fill ="Signature (Aetiology)") +
  theme_minimal(base_size = 11) +
  scale_fill_manual(values = sig_colours_cosmicv2) +
  theme(axis.text.x=element_blank(),
        strip.text.x = element_text(size = 6),
        panel.grid.major.y = element_blank(),
        panel.grid.major.x = element_blank(),
        legend.position = "right")
```

# De novo mutational signature comparison to Cosmic v2
  
Cosine similarity between 6 de novo signatures (SigProfiler) & Cosmic v2 30 signatures.  
```{r cosine_denovo_cosmicv2, fig.width=8,fig.height=6}
de_novo_sigs_df <- read_tsv("./data/other/Signature_6.tsv", col_types = cols()) 

de_novo_sigs <- de_novo_sigs_df %>%
  mutate(base1 = substr(Subtype, start = 1, stop = 1), base3 = substr(Subtype, start = 3, stop = 3)) %>%
  mutate(context = paste0(base1,"[",Type,"]",base3)) %>%
  column_to_rownames(var = "context") %>%
  select(-c(base1,base3,Type,Subtype))

cos_sim_sig_cosmicv2 <- cos_sim_matrix(cosmic_t, de_novo_sigs)

row_order_v2 <-  rev(colnames(cosmic_t))


# Plot heatmap with specified signature order
Heatmap(cos_sim_sig_cosmicv2, col= col_fun_cosine, cluster_columns = FALSE,
        cluster_rows = FALSE, clustering_method_rows = "single",
        row_order = row_order_v2,
        row_names_gp = gpar(fontsize = 6),column_names_gp = gpar(fontsize = 7),
        heatmap_legend_param = list(title = "cosine\nvalue"),
        cell_fun = function(j, i, x, y, width, height, fill) 
          if(cos_sim_sig_cosmicv2[i, j] > 0.65) {
            grid.text(sprintf("%.2f", cos_sim_sig_cosmicv2[i, j]), x, y, gp = gpar(fontsize = 6, col = "white"))
          } else {
            grid.text(sprintf("%.2f", cos_sim_sig_cosmicv2[i, j]), x, y, gp = gpar(fontsize = 6))
          }
)

```

```{r denovo_sig_samples, message=FALSE}

de_novo_sigs_assignment <- read_csv("./data/other/Distribution_Analysis_6_Signatures.csv", col_types = cols()) %>%
  select(-ends_with("_Exposure")) %>%
  gather(key = Sig, value = Probability, -testSampleLabel) %>%
  mutate(Sig = str_remove(Sig,"_Percent"),
         Count = Probability/100) %>%
  right_join(maf_ghcc %>% select(aliquot_id,case_submitter_id) %>% distinct(),
            by=c("testSampleLabel"="aliquot_id")) %>%
  left_join(sample_list) %>%
  dplyr::rename(Sample = TCGA_cases) %>%
  mutate(Sig_full = case_when(Sig == "Sig_A" ~ "Sig A (POLE mut, Sig10)",
                              Sig == "Sig_B" ~ "Sig B (Age, Sig1)",
                              Sig == "Sig_C" ~ "Sig C (MSI, Sig20)",
                              Sig == "Sig_D" ~ "Sig D (MSI, Sig26)",
                              Sig == "Sig_E" ~ "Sig E (POLE mut & MSI, Sig14)",
                              Sig == "Sig_F" ~ "Sig F (APOBEC, Sig2)")) %>%
  mutate(Sig_full = factor(Sig_full, 
                           levels = rev(sig_list_denovo))) %>%
  filter(Sample %in% germ_vars_path$Sample)


p1 <- de_novo_sigs_assignment %>%
  filter(Sample %in% unique(germ_vars_path$Sample)) %>%
  mutate(Sample = factor(Sample, levels = unique(germ_vars_path$Sample))) %>%
  ggplot(.) +
  geom_bar(aes(x= Sample, y = Count, fill = Sig_full), stat = "identity", position = "stack") + 
  labs(x="",y="Proportion",fill ="De novo Signature (Aetiology)") +
  theme_minimal(base_size = 8) +
  geom_hline(yintercept = 0.25, linetype = "dashed") + 
  scale_fill_manual(values = sig_colours_denovo) +
  guides(fill = guide_legend(ncol = 1)) +
  theme(axis.text.x=element_blank(),
        strip.text.x = element_text(size = 6),
        panel.grid.minor = element_blank(),
        #panel.grid.major.y = element_blank(),
        panel.grid.major.x = element_blank(),
        legend.position = "right")

p2 <- germ_vars_path %>%
  filter(Sample %in% p1$data$Sample) %>%
  filter(Gene %in% c("MSH2","MSH6","PMS2")) %>%
  mutate(Sample = factor(Sample, levels = unique(germ_vars_path$Sample))) %>%
  ggplot(.) +
  geom_tile(aes(y=Gene,x=Sample,fill=Type, colour = Enriched, width = 0.8, height = 0.8), size = 1)+
  scale_fill_manual(values = c("WT"="#d6d6d6","Pathogenic"="#1e6dc0")) + 
  #scale_fill_manual(values= mut_colours) +
  scale_colour_manual(values = c("yes"="#E8294A","white")) +
  guides(color = guide_legend(ncol = 1,
                              override.aes = list(fill=NA))) +
  theme_minimal(base_size = 11) + labs(x = "", y = "", colour = "Enriched in tumour",
                                       fill = "Germline Status") +
  theme(axis.text.x=element_text(angle = 90, hjust = 1 , vjust = 0.5),
        panel.grid.major.y = element_blank(),
        panel.grid.major.x = element_blank(),
        legend.position = "right",
        strip.text.y = element_blank())

plot_grid(p1, p2, align = "v",axis = 'lr', nrow = 2, rel_heights = c(0.5,0.4))

ggsave("./figures/denovo_sig_samples.pdf", width = 8, height = 6)

```

